### ISSUE TITLE
Strange behavior at first call and after 5s

### ISSUE THREAD

**Issue by cabrinoob** [fastify/fastify#1022]
#### Context

* *node version*: 10.0.0
* *fastify version*: ^1.6.0

Hi,

I'am using fastify through a small nestjs project. I have a single GET REST method. 

Using Fastify, the first call takes 500ms at minimum and if I continue calling the same method, the response time falls arround 4ms.

BUT, if I wait 5 seconds between 2 calls, the second call replies after 500ms again.

I do not see this problem using Express.

Anyone else experienced this behavior?

Thank you

**Comment by mcollina** [comment]
Can you please include a snippet to reproduce your problem?

**Comment by delvedor** [comment]
Closing due to inactivity, feel free to reopen!

**Comment by cabrinoob** [comment]
Ok, I don't know how to reopen it, so I post here : 

```
const fastify = require('fastify')()

// Declare a route
fastify.get('/', async (request, reply) => {
  return { hello: 'world' }
})

// Run the server!
const start = async () => {
  try {
    await fastify.listen(3000)
    fastify.log.info(`server listening on ${fastify.server.address().port}`)
  } catch (err) {
    fastify.log.error(err)
    process.exit(1)
  }
}
start()
```

I just start this basic doc sample on my local computer (Windows 10, nodeJS v10.0.0)

Then I simply open postman and call `http://localhost:3000`

The very first call takes 500+ ms
Immediates next calls take 2ms

And then, if you wait 5 seconds, the next call takes 500+ ms (as if it was the first one)

**Comment by delvedor** [comment]
I'm not able to reproduce this under macOS. @StarpTech can you try this?

**Comment by cabrinoob** [comment]
I've tried on another teammate computer, also on windows10 but under nodeJS v8.x and he has the same behavior (with fiddler4 client).

**Comment by mcollina** [comment]
I cannot reproduce this on Windows either. Might it be some antivirus/firewall? It might also be a client issue, have you tried with postman?

We disconnect keepalive connections after 5s.

**Comment by cabrinoob** [comment]
I have the same behavior with fiddler4 and postman. I also disabled my firewall. Still the same behavior.

So it seems to be the 5s timeout on keep-alive connections that is the cause of this problem?

**Comment by mcollina** [comment]
@cabrinoob do you experience the same thing with Chrome? Can you check with Node 10.6.0 if it has the same issue? We inherit that behavior from Node.js core, see https://nodejs.org/api/http.html#http_server_keepalivetimeout.

Does Node 8 has the same problem?

**Comment by cabrinoob** [comment]
@mcollina : I do not experience exactly the same behavior with Chrome. If I wait 6 or 7 seconds between 2 calls, it responds fast (about 20ms in the console). But if I wait more than 10s it responds in 300+ ms.

So, with chrome I do not exactly see the same thing.

After that, I tried with nodejs 10.6.0 and I did not see any change : same behavior with fiddler4/postman than before, and same behavior with chrome than before.

**Comment by mcollina** [comment]
Does Node 8 has the same issue?

**Comment by justintime4tea** [comment]
I am experiencing this as well with node 8 (my own project, similar scope):
 
node v8.9.0
Windows 10 both on WSL and native. 

Going to test later under Linux because as long as it doesn't happen there it's a moot point for me.

**Comment by mcollina** [comment]
Does this happens with https://github.com/fastify/benchmarks/blob/master/benchmarks/bare.js?

I cannot reproduce this at all.

**Comment by cabrinoob** [comment]
@mcollina : using `bare.js` I cannot see the problem. I have some fluctuations between 2ms and 20ms, but nothing compared to the gap 2ms - 500ms observed with fastify.

**Comment by mcollina** [comment]
Unfortunately I cannot reproduce on Windows, Linux or Mac OS X. It seems very much system dependent.

I'll be happy to review a PR that fixes this problem (if it's a problem with fastify).

**Comment by justintime4tea** [comment]
I cam reproduce this issue with the minimal fastify example posted here by @cabrinoob but the issue is not reproducible by the bare.js posted by mcollina. This is the case whether I'm using WSL or Windows (have not tested under Linux, doing this next)

**Comment by delvedor** [comment]
Can you reproduce it with https://github.com/fastify/benchmarks/blob/master/benchmarks/fastify.js?
(no async await)

**Comment by cabrinoob** [comment]
@delvedor : Yes, I reproduce the 500ms lag with your code.

**Comment by StarpTech** [comment]
@delvedor I will try to reproduce it tonight and report.

**Comment by mcollina** [comment]
I can reproduce the problem. If I open `http://localhost:3000` I get the delay. If I open `http://127.0.0.1:3000` I do not get the delay.

@justintime4tea @cabrinoob 

Moreover I can reproduce it with plain core: 

```
'use strict'

const server = require('http').createServer(function (req, res) {
  res.setHeader('content-type', 'application/json')
  res.end(JSON.stringify({ hello: 'world' }))
})

server.listen(3000, '127.0.0.1')
```

Then opening `http://localhost:3000` will have the delay.

It seems that _just_ adding that 2nd property in `.listen()` slows down Windows for some reason.
This looks like a bug on Node.js core on Windows.

Side note, this is triggered by passing in _any_ value.

cc @jsumners

**Comment by justintime4tea** [comment]
@mcollina You are awesome! Thank you so much for the feedback on this! We're using Linux in production so it's good to know we won't be seeing this in production. Good ol' Winderz >_<

**Comment by cabrinoob** [comment]
Ok, great!
For the fun, I've just tested with `server.listen(3000)` (without forcing the 127.0.0.1) and make requests with postman on 127.0.0.1 and it works like a charm.

switching on localhost brings up the problem.

**Comment by mcollina** [comment]
Confirmed it's a WIndows problem. The following does not cause the slowdown:

```js
'use strict'

const server = require('http').createServer(function (req, res) {
  res.setHeader('content-type', 'application/json')
  res.end(JSON.stringify({ hello: 'world' }))
})

server.listen(3000, '::1')
```

This is happening because you have IPv6 enabled on your machine, and `localhost` is first resolving to `::1` (`localhost` in IPv6), and _then_ to `127.0.0.1`. On Mac OS X, `localhost` resolves to `127.0.0.1`. I assume this is the same for Linux.

As for solving this bug, I see two possible ways:

1. lookup what `localhost` is, and then call `.listen()` on that address. Some docs will need to be updated, as we will not be bound to IPv4 anymore. Considering this would not affect production deployments, I think we can ship this in a semver-minor.

2. do not listen a specific address at all, but rather to all specific interfaces. This reduces some base security as it will expose the server over the network.

I'm for 1. @fastify/fastify what do you think?

**Comment by mcollina** [comment]
The 1. approach is implemented in https://github.com/fastify/fastify/pull/1047.

**Comment by jsumners** [comment]
As I say in #1047, I think the solution is:

3. Add a note to the documentation explaining that "localhost" usually references multiple addresses which can cause name resolution delays. Fastify explicitly binds to `127.0.0.1`. Thus, to avoid name resolution delays, requests should be made to `http://127.0.0.1...` instead of `http://localhost...`.

**Comment by xnerhu** [comment]
Same with Node `v12.16.3` on Windows.

**Comment by mcollina** [comment]
Very likely it’s a different bug or you are doing something odd on your end.

**Comment by StarpTech** [comment]
Hi, awesome to land here 😄 I research for hours. I can reproduce it with Node.js 14.17.2 and the latest 16. I use WSL2.

Listen to `::1` fix it but `127.0.0.1` or `localhost` produce a delay for more than 200ms in our servers.
As @jsumners mentioned always request `127.0.0.1` and it works.
